brian2/codegen/runtime/numpy_rt/templates/spikegenerator.py_

```diff
@@ -1,12 +1,41 @@
-{# USES_VARIABLES { _spikespace, t, dt, neuron_index, spike_time } #}
+{# USES_VARIABLES { _spikespace, t, dt, neuron_index, spike_time, period, _lastindex } #}
 import numpy as np
 
 # TODO: We don't deal with more than one spike per neuron yet
-# Spikes since the preceding time step
-start = np.searchsorted({{spike_time}}, t-dt, side='right')
-stop = np.searchsorted({{spike_time}}, t, side='right')
-_indices = {{neuron_index}}[start:stop]
 
-_n_spikes = len(_indices)
+padding_before = t % period
+padding_after  = (t + dt) % period
+_n_spikes      = 0
+epsilon        = 1e-3*dt
+
+# We need some precomputed values that will be used during looping
+not_first_spike = {{_lastindex}}[0] > 0
+not_end_period  = abs(padding_after) > (dt - epsilon)
+
+# If there is a periodicity in the SpikeGenerator, we need to reset the lastindex 
+# when all spikes have been played and at the end of the period
+if not_first_spike and ({{spike_time}}[{{_lastindex}}[0] - 1] > padding_before):
+    {{_lastindex}}[0] = 0
+
+for _spike_time in {{spike_time}}[{{_lastindex}}[0]:]:
+    test = (_spike_time > padding_before) or (abs(_spike_time - padding_before) < epsilon)
+    if test:
+        break
+    {{_lastindex}}[0] += 1
+
+for _spike_time in {{spike_time}}[{{_lastindex}}[0]:]:
+    if not_end_period:
+        test = (_spike_time > padding_after) or (abs(_spike_time - padding_after) < epsilon)
+    else:
+        # If we are in the last timestep before the end of the period, we remove the first part of the
+        # test, because padding will be 0
+        test = abs(_spike_time - padding_after) < epsilon
+    if test:
+        break
+    _n_spikes += 1
+
+_indices = {{neuron_index}}[{{_lastindex}}[0]:{{_lastindex}}[0]+_n_spikes]
+
```

Comment:
Actually, I ran some crude benchmarks and for big populations/many spikes it is considerably worse now... In fact, in the previous implementation I don't see any scaling with the number of spikes (it should be logarithmic because of the binary search). I guess the above code should be significantly faster if expressed in a vectorized way instead of using a loop. Obviously it would be nice if we don't decrease performance with the new, "faster" approach...